File: packages/earthsci-toolkit/src/geometry-fixtures.test.ts
/**
 * Schema-validation conformance tests for the additive M4 geometry-kernel
 * fixtures (bead ess-my4.4.2; RFC semiring-faq-unified-ir §8.1 / §A.8;
 * CONFORMANCE_SPEC.md §5.8). Exercises the additive schema deltas — the
 * `intersect_polygon` leaf op, its required `manifold` flag (planar /
 * spherical / geodesic), and the bin-Skolem spatial-join representation
 * composed from the existing `floor` / `skolem` / `join.on` ops — through the
 * TypeScript binding's validate + parse/serialize path. The TS binding does no
 * numeric polygon clipping, so this is validation / round-trip only; the
 * tolerance-based cross-binding clip conformance is asserted by the
 * Julia / Rust / Python evaluator suites (ess-my4.4.3/.4/.11/.12, gated by
 * §5.8). Mirrors aggregate-fixtures.test.ts.
 */
import { describe, it, expect } from 'vitest';
import { readFileSync, readdirSync } from 'fs';
import { join, basename } from 'path';
import { load, save, validate, validateSchema } from './index.js';

const testsDir = join(__dirname, '../../../tests');

function esmFilesIn(dir: string): string[] {
  return readdirSync(dir)
    .filter((e) => e.endsWith('.esm'))
    .map((e) => join(dir, e))
    .sort();
}

describe('Geometry kernel (intersect_polygon) fixtures', () => {
  describe('tests/valid/geometry', () => {
    const validFiles = esmFilesIn(join(testsDir, 'valid', 'geometry'));

    it('has fixtures to test', () => {
      expect(validFiles.length).toBeGreaterThan(0);
    });

    it.each(validFiles)('validates and round-trips %s', (filePath) => {
      const content = readFileSync(filePath, 'utf-8');

      // Schema-valid: the additive intersect_polygon op + manifold flag, the
      // derived-index-set ring (from_faq), and the floor/skolem/join.on
      // bin-Skolem composition all satisfy the embedded JSON schema.
      expect(validateSchema(JSON.parse(content))).toHaveLength(0);

      // Structurally valid too: the structural pass is intersect_polygon-aware
      // — the clip leaf is walked like any other op (its two polygon operands
      // are ordinary sub-expression references), and polygon_area / the
      // bin-Skolem join ride the existing aggregate-aware reference + balance
      // machinery, so nothing is flagged as undefined.
      const result = validate(content);
      expect(result.schema_errors).toHaveLength(0);
      expect(result.structural_errors).toHaveLength(0);
      expect(result.is_valid).toBe(true);

      // parse -> serialize -> parse is a fixed point on the typed view
      // (the `manifold` flag survives the typed round-trip).
      const original = load(content);
      const reloaded = load(save(original));
      expect(reloaded).toEqual(original);
    });
  });

  describe('tests/invalid/geometry', () => {
    const invalidFiles = esmFilesIn(join(testsDir, 'invalid', 'geometry'));

    it('has fixtures to test', () => {
      expect(invalidFiles.length).toBeGreaterThan(0);
    });

    // Each geometry invalid fixture is a pure SCHEMA violation isolated to the
    // intersect_polygon node — a missing `manifold`, a third operand (the op is
    // strictly binary), or a `manifold` outside the closed enum — so the schema
    // validator alone rejects them.
    it.each(invalidFiles)('rejects %s', (filePath) => {
      const parsed = JSON.parse(readFileSync(filePath, 'utf-8'));
      expect(validateSchema(parsed).length).toBeGreaterThan(0);
    });
  });
});

File: packages/esm-format-go/pkg/esm/geometry_fixtures_test.go
package esm

import (
	"path/filepath"
	"testing"
)

// TestGeometryValidFixtures asserts every tests/valid/geometry/*.esm fixture
// parses and schema-validates cleanly through the Go loader. These exercise the
// additive M4 geometry-kernel schema deltas (bead ess-my4.4.2; RFC
// semiring-faq-unified-ir §8.1 / §A.8): the `intersect_polygon` leaf op, its
// required `manifold` flag, the clipped overlap ring exposed as a kind:"derived"
// index set (the ring rides the M3 value-invention machinery), and the
// bin-Skolem spatial-join representation composed from the existing
// `floor`/`skolem`/`join.on` ops. Validation/round-trip only — the Go binding
// does no polygon clipping; the tolerance-based clip conformance is the
// evaluator suites' (CONFORMANCE_SPEC.md §5.8). Mirrors
// TestAggregateValidFixtures.
func TestGeometryValidFixtures(t *testing.T) {
	repoRoot, err := filepath.Abs(filepath.Join("..", "..", "..", ".."))
	if err != nil {
		t.Fatalf("resolve repo root: %v", err)
	}
	pattern := filepath.Join(repoRoot, "tests", "valid", "geometry", "*.esm")
	files, err := filepath.Glob(pattern)
	if err != nil {
		t.Fatalf("glob %s: %v", pattern, err)
	}
	if len(files) == 0 {
		t.Fatalf("no .esm fixtures matched %s", pattern)
	}
	for _, path := range files {
		name := filepath.Base(path)
		t.Run(name, func(t *testing.T) {
			if _, err := Load(path); err != nil {
				t.Fatalf("expected %s to validate, got error: %v", name, err)
			}
		})
	}
}

// TestGeometryInvalidFixtures asserts every tests/invalid/geometry/*.esm fixture
// is handled correctly. Each is a pure schema violation isolated to the
// intersect_polygon node — a missing `manifold` (it is required, no default), a
// third operand (the clip is strictly binary), or a `manifold` outside the
// closed {planar, spherical, geodesic} enum — so Load returns a non-nil error
// at schema-validation time. A resolver_only geometry fixture (none today) would
// be schema-valid and accepted, mirroring TestAggregateInvalidFixtures.
func TestGeometryInvalidFixtures(t *testing.T) {
	repoRoot, err := filepath.Abs(filepath.Join("..", "..", "..", ".."))
	if err != nil {
		t.Fatalf("resolve repo root: %v", err)
	}
	resolverOnly := resolverOnlyInvalidFixtures(t, repoRoot)
	pattern := filepath.Join(repoRoot, "tests", "invalid", "geometry", "*.esm")
	files, err := filepath.Glob(pattern)
	if err != nil {
		t.Fatalf("glob %s: %v", pattern, err)
	}
	if len(files) == 0 {
		t.Fatalf("no .esm fixtures matched %s", pattern)
	}
	for _, path := range files {
		name := filepath.Base(path)
		t.Run(name, func(t *testing.T) {
			if resolverOnly[name] {
				if _, err := Load(path); err != nil {
					t.Fatalf("resolver-only fixture %s must pass schema validation, got error: %v", name, err)
				}
				return
			}
			if _, err := Load(path); err == nil {
				t.Fatalf("expected %s to be rejected, but it validated", name)
			}
		})
	}
}
